hazelcast/src/main/java/com/hazelcast/client/impl/protocol/ClientMessage.java

```diff
@@ -16,367 +16,100 @@
- * </pre>
  */
-public class ClientMessage
-        extends MessageFlyweight
-        implements OutboundFrame {
+@SuppressWarnings("checkstyle:MagicNumber")
+@BinaryInterface
+public final class ClientMessage extends LinkedList<ClientMessage.Frame> implements OutboundFrame {
 
-    /**
-     * Current protocol version
-     */
-    public static final short VERSION = 1;
-
-    /**
-     * Begin Flag
-     */
-    public static final short BEGIN_FLAG = 0x80;
-
-    /**
-     * End Flag
-     */
-    public static final short END_FLAG = 0x40;
-
-    /**
-     * Begin and End Flags
-     */
-    public static final short BEGIN_AND_END_FLAGS = (short) (BEGIN_FLAG | END_FLAG);
+    public static final int TYPE_FIELD_OFFSET = 0;
+    public static final int CORRELATION_ID_FIELD_OFFSET = TYPE_FIELD_OFFSET + Bits.SHORT_SIZE_IN_BYTES;
 
-    /**
-     * Listener Event Flag
-     */
-    public static final short LISTENER_EVENT_FLAG = 0x01;
-
-    /**
-     * ClientMessage Fixed Header size in bytes
-     */
-    public static final int HEADER_SIZE;
+    //offset valid for fragmentation frames only
+    public static final int FRAGMENTATION_ID_OFFSET = 0;
 
-    private static final String PROP_HAZELCAST_PROTOCOL_UNSAFE = "hazelcast.protocol.unsafe.enabled";
-    private static final boolean USE_UNSAFE = Boolean.getBoolean(PROP_HAZELCAST_PROTOCOL_UNSAFE);
+    //optional fixed partition id field offset
+    public static final int PARTITION_ID_FIELD_OFFSET = CORRELATION_ID_FIELD_OFFSET + Bits.LONG_SIZE_IN_BYTES;
 
-    private static final int FRAME_LENGTH_FIELD_OFFSET = 0;
-    private static final int VERSION_FIELD_OFFSET = FRAME_LENGTH_FIELD_OFFSET + Bits.INT_SIZE_IN_BYTES;
-    private static final int FLAGS_FIELD_OFFSET = VERSION_FIELD_OFFSET + Bits.BYTE_SIZE_IN_BYTES;
-    private static final int TYPE_FIELD_OFFSET = FLAGS_FIELD_OFFSET + Bits.BYTE_SIZE_IN_BYTES;
-    private static final int CORRELATION_ID_FIELD_OFFSET = TYPE_FIELD_OFFSET + Bits.SHORT_SIZE_IN_BYTES;
-    private static final int PARTITION_ID_FIELD_OFFSET = CORRELATION_ID_FIELD_OFFSET + Bits.LONG_SIZE_IN_BYTES;
-    private static final int DATA_OFFSET_FIELD_OFFSET = PARTITION_ID_FIELD_OFFSET + Bits.INT_SIZE_IN_BYTES;
+    public static final int DEFAULT_FLAGS = 0;
+    public static final int BEGIN_FRAGMENT = 1 << 15;
+    public static final int END_FRAGMENT = 1 << 14;
+    public static final int UNFRAGMENTED_MESSAGE = BEGIN_FRAGMENT | END_FRAGMENT;
+    public static final int FINAL = 1 << 13;
+    public static final int BEGIN_DATA_STRUCTURE = 1 << 12;
+    public static final int END_DATA_STRUCTURE = 1 << 11;
+    public static final int IS_NULL = 1 << 10;
+    public static final int IS_EVENT = 1 << 9;
 
-    static {
+    //frame length + flags
+    public static final int SIZE_OF_FRAMELENGHT_AND_FLAGS = Bits.INT_SIZE_IN_BYTES + Bits.SHORT_SIZE_IN_BYTES;
+    public static final Frame NULL_FRAME = new Frame(new byte[0], IS_NULL);
```

Comment:
I have made null illegal on purpose. If I don't we need to differentiate null and 0 sized byte array, which is not easy because size field is not the only size of the byte[]. 
Making `null` illegal was making the code a lot simpler.